projs/shadow/shadow-engine/core/src/core/module-manager-v2.cpp

```diff
@@ -0,0 +1,216 @@
+                return a.id == i.descriptor.assembly && a.lib != nullptr;
+            })) {
+                spdlog::debug("✅ Assembly \"{0}\" is already loaded", i.descriptor.assembly);
+            } else {
+                spdlog::info("➕ Loading assembly \"{0}\"", i.descriptor.assembly);
+                this->LoadAssembly(i.descriptor.assembly);
+            }
+
+            this->LoadModule(i);
+            if (i.enabled)
+                spdlog::info("Module {0}({1}) created", i.module->GetName(), i.descriptor.id);
+        }
+
+
+
+        //PreInit
+        spdlog::info("Running PreInit");
+        for (auto &holder: this->modules) {
+            if (holder.enabled) {
+                holder.module->PreInit();
+            }
+        }
+
+        spdlog::info("Running PreInit");
+        for (auto &holder: this->modules) {
+            if (holder.enabled) {
+                holder.module->Init();
+            }
+        }
+
+        //Sort
+
+        this->finalized = true;
+    }
+
+    void ModuleManager::LoadAssembly(const std::string &id) {
+        auto assembly = std::find_if(ITERATE(this->assemblies), [&](const auto &item) {
+            return item.id == id;
+        });
+        if (assembly == this->assemblies.end()) {
+            spdlog::error("No such assembly: {0}", id);
+            return;
+        }
+
+
+        dylib *dllptr;
+
+        try {
+            dllptr = new dylib("./", assembly->path);
+        }
+        catch (std::exception &e) {
+            spdlog::error(e.what());
+            exit(1);
+        }
+
+        assembly->lib = dllptr;
+    }
+
+    void ModuleManager::dfs(const ModuleHolder &moduleHolder, std::vector<ModuleHolder> &sorted) {
+        for (auto u: moduleHolder.descriptor.dependencies) {
+            if (!std::any_of(ITERATE(sorted), ModulePredicate(u)) && u != moduleHolder.descriptor.id) {
+                auto it = std::find_if(ITERATE(this->modules), ModulePredicate(u));
+
+                if (it != modules.end())
+                    dfs(*it, sorted);
+                else
+                    spdlog::info("Module {0} is missing, required by {1}", u, moduleHolder.descriptor.id);
+
+            }
+        }
+        sorted.push_back(moduleHolder);
+    }
+
+    void ModuleManager::SortModules() {
+        std::vector<ModuleHolder> sorted;
+        sorted.clear();
+
+        for (auto i: this->modules) {
+            if (!std::any_of(ITERATE(sorted), ModulePredicate(i.descriptor.id)))
+                dfs(i, sorted);
+        }
+
+        this->modules = sorted;
+
+    }
+
+    void ModuleManager::LoadModulesFromAssembly(const std::string &id) {
+        LoadAssembly(id);
+
+        auto assembly = GetAssembly(id);
+        const std::string &symbolName = "assembly_entry";
+
+        if (!assembly.lib->has_symbol(symbolName)) {
+            spdlog::error("❌ Could not find the entry for assembly \"{0}\"({1})", id, assembly.path.c_str());
```

Comment:
Encoding is fun! Please turn this into a Unicode literal, so that my editor doesn't fuck this up.